script/ebooks

```diff
@@ -0,0 +1,196 @@
+  .use(replaceAsides)
+  .use(replaceCheckboxes)
+  .use(fixImages)
+  .use(downloadImages)
+  .use(stringify);
+
+buildBooks([
+  '_articles/how-to-contribute.md',
+  '_articles/starting-a-project.md',
+  '_articles/finding-users.md',
+  '_articles/building-community.md',
+  '_articles/best-practices.md',
+  '_articles/leadership-and-governance.md',
+  '_articles/getting-paid.md',
+  '_articles/code-of-conduct.md',
+  '_articles/metrics.md',
+  '_articles/legal.md',
+]);
+
+function buildBooks(files) {
+  const out = 'assets/ebooks/open-source-guide';
+  const tempFile = '_ebook.md';
+
+  mkdirp.sync('assets/ebooks');
+  mkdirp.sync('.asset-downloads');
+
+  let source = files.map(readFile).join('\n\n');
+  const { text, promises } = downloadImages(source);
+  fs.writeFileSync(tempFile, text);
+
+  promises
+    .then(() => markdownToEpub(tempFile, `${out}.epub`))
+    .then(() => epubToMobi(`${out}.epub`, `${out}.mobi`))
+    .then(() => fs.unlinkSync(tempFile))
+    .catch(err => console.error(err));
+}
+
+function readFile(file) {
+  const md = fs.readFileSync(file, 'utf8');
+
+  // Replace frontmatter with chapter title
+  const { attributes, body } = fastmatter(md);
+  let text = `# ${attributes.title}\n\n${body}`;
+
+  return processor.process(text).contents;
+}
+
+function markdownToEpub(mdFile, epubFile) {
+  return new Promise((resolve, reject) => {
+    console.log(`Building ${epubFile}...`);
+
+    const manifestFile = '_ebook.json';
+    fs.writeFileSync(manifestFile, JSON.stringify({
+      tocDepth: 1,
+      title: 'Open Source Guides',
+      subtitle: 'Open source software is made by people just like you. Learn how to launch and grow your project.',
+      author: 'GitHub',
+      publisher: 'GitHub',
+      rights: 'CC-BY-4.0',
+      date: (new Date()).toISOString().slice(0, 10),
+      contents: path.resolve(mdFile),
+    }));
+
+    ebrew.generate(manifestFile, epubFile, err => {
+      fs.unlinkSync(manifestFile);
+      if (err) {
+        reject(err.stack || err);
+      } else {
+        resolve();
+      }
+    });
+  });
+}
+
+function epubToMobi(epubFile, mobiFile) {
+  return new Promise((resolve, reject) => {
+    console.log(`Building ${mobiFile}...`);
+
+    const kindlegen = require.resolve('kindlegen/bin/kindlegen');
+    exec(`${kindlegen} ${epubFile} -c2 -verbose -o ${path.basename(mobiFile)}`, (err, stdout) => {
+      if (err && stdout.includes('MOBI file could not be generated because of errors!')) {
+        reject(stdout);
+      } else {
+        resolve();
+      }
+    });
+  });
+}
+
+function downloadImages(text) {
+  function image(processor) {
+    return ast => visit(ast, 'image', node => {
+      if (node.url.startsWith('http')) {
+        const url = node.url;
+        const filename = slug(path.basename(node.url));
+        const filepath = `./.asset-downloads/${filename}.jpg`
+        node.url = filepath;
+        if (fs.existsSync(filepath)) {
+          promises.push(Promise.resolve());
```

Comment:
@sapegin Why is it needed to add a resolving promise here?

Would `if (!fs.existsSync(filepath)) { download() ...` not be simpler?